### PR TITLE
fee invariance and tests

### PR DESCRIPTION
**Pull request by marta-lokhova**
Addresses the first two bullets in https://github.com/stellar/stellar-core/issues/4007

- Reject malformed soroban txs as early as possible. Harden resource fee access (can't be negative)
- Force fee checks by disallowing negative inclusion fee calculation
- Test: negative resource and inclusion fees
- Test: insufficient balance to cover full fee (with and without replace-by-fee tx discount)

### CODE REVIEW

File: src/transactions/FeeBumpTransactionFrame.cpp [hunk 1]
```diff
@@ -321,11 +321,26 @@ FeeBumpTransactionFrame::getFullFee() const
     return mEnvelope.feeBump().tx.fee;
 }
 
+int64
+FeeBumpTransactionFrame::declaredSorobanResourceFee() const
```
Comment (dmkozh) [inline src/transactions/FeeBumpTransactionFrame.cpp:325]: Where are we using this?

File: src/transactions/TransactionFrame.cpp [hunk 2]
```diff
@@ -1352,6 +1355,24 @@ TransactionFrame::processFeeSeqNum(AbstractLedgerTxn& ltx,
     }
 }
 
+bool
+TransactionFrame::isValidStructure() const
```
Comment (dmkozh) [inline src/transactions/TransactionFrame.cpp:1359]: nit: I think `isValidStructure` is a bit misleading given that this is a required invariant for fee-related processing only. So maybe rename this to reflect that better, like `isValidForFeeProcessing` or something like that. Maybe could add some negative fee protection in fee bump envelope as well. Up to you though - if you want to put non-fee-related checks here (but it seems like that would dilute the validation logic a bit too much IMHO)

File: src/transactions/TransactionFrame.cpp [hunk 3]
```diff
@@ -1629,10 +1650,13 @@ TransactionFrame::applyOperations(SignatureChecker& signatureChecker,
             if (protocolVersionStartsFrom(ledgerVersion,
                                           SOROBAN_PROTOCOL_VERSION))
             {
-                // If transaction fails, we don't charge for any
-                // refundable resources.
-                mFeeRefund = declaredSorobanResourceFee() -
-                             mSorobanResourceFee->non_refundable_fee;
+                if (isSoroban())
```
Comment (dmkozh) [inline src/transactions/TransactionFrame.cpp:1653]: `isSoroban` should be moved to the check above as we don't want to add diagnostic events to the classic txs. It's ok though to postpone this to my change since I'm regenerating the tx meta anyway, though there is still this weird breakage in my PR, so maybe it's better to regenerate it for your (smaller) PR.

File: src/transactions/test/TxEnvelopeTests.cpp [hunk 4]
```diff
@@ -2513,13 +2513,6 @@ TEST_CASE("soroban transaction validation", "[tx][envelope][soroban]")
         }
     };
 
-    SECTION("no soroban extension")
```
Comment (dmkozh) [inline src/transactions/test/TxEnvelopeTests.cpp:2516]: Why is this removed?

Comment (marta-lokhova) [reply src/transactions/test/TxEnvelopeTests.cpp:2516]: We shouldn't allow malformed transactions to get this far (to check valid), so the extension check moved up in the flow, and now we reject malformed transactions at the tx queue stage, hence I moved this test case to `"Soroban TransactionQueue limits"`

File: src/transactions/FeeBumpTransactionFrame.cpp (hunk 1 above)
Comment (marta-lokhova) [reply src/transactions/FeeBumpTransactionFrame.cpp:325]: in `"negative inclusion fee"` test case. it's not used much, but it seems harmless to have this helper, so callers don't need to know the internals of fee bumps to extract the inner tx's resource fee.

File: src/herder/TransactionQueue.cpp [hunk 5]
```diff
@@ -245,9 +245,11 @@ TransactionQueue::canAdd(TransactionFrameBasePtr tx,
             mApp.getConfig());
     }
 
+    // newFullFee, cast from uint32_t, so it can't be negative
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/TransactionQueue.cpp:248]: good catch, thanks, I looked at the wrong fee field it seems

File: src/herder/TransactionQueue.cpp [hunk 6]
```diff
@@ -245,9 +245,11 @@ TransactionQueue::canAdd(TransactionFrameBasePtr tx,
...
     int64_t newFullFee = tx->getFullFee();
-    if (tx->getInclusionFee() < 0 || newFullFee < 0)
+    if (tx->getInclusionFee() < 0)
     {
+        tx->getResult().result.code(txSOROBAN_INVALID);
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/TransactionQueue.cpp:252]: switched to `txMALFORMED`

File: src/transactions/test/TxEnvelopeTests.cpp (hunk 4 above)
Comment (dmkozh) [reply src/transactions/test/TxEnvelopeTests.cpp:2516]: > We shouldn't allow malformed transactions to get this far (to check valid)

I think that's a wrong assumption. When we're getting a tx set, we need to call `checkValid` on every transaction, so a malformed tx set would now crash Core apparently (we probably don't have a test for this particular error condition). In general, I think it's not correct to make any assumptions on the flows that are external to tx frame - if it has a public function, it should be well-behaved in all the appropriate contexts.

File: src/transactions/FeeBumpTransactionFrame.cpp (hunk 1 above)
Comment (dmkozh) [reply src/transactions/FeeBumpTransactionFrame.cpp:325]: If that's test-only, then I would suggest to make it public (and part of the interface) only for tests - so far there wasn't an intent/need to access soroban resource fee outside of the tx frame itself.

File: src/herder/test/TransactionQueueTests.cpp [hunk 7]
```diff
@@ -2167,14 +2208,26 @@ TEST_CASE("transaction queue with fee-bump", "[herder][transactionqueue]")
             TransactionQueueTest test{queue};
             auto tx1 = transaction(*app, account1, 1, 1, 100, 1, isSoroban);
             uint32_t discount = tx1->getFullFee() - tx1->getInclusionFee();
+            auto newInclusionToPay = 200;
+            if (isSoroban)
+            {
+                // In case of Soroban, provide additional discount to test the
+                // case where inclusion fee is less than balance, but total fee
+                // is not.
+                discount += newInclusionToPay;
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/test/TransactionQueueTests.cpp:2217]: What do you mean here exactly? The test was already testing the right scenarios (not enough balance to cover the fee), but because it used only classic transactions, it did not check that there's enough balance to cover full fee, not just inclusion. So the setup changed a bit to make sure with Soroban, the account has enough balance to cover inclusion fee, but not full fee.

[the remark this answers is not in the document]
Comment (dmkozh) [reply src/herder/test/TransactionQueueTests.cpp:2217]: In general I'd expect that if test setup changes, then either expectations, or covered logic changes. I guess it's ok that it still passes, but maybe there is some additional assertion that would fail before this change, but would pass after adding it.

[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/test/TransactionQueueTests.cpp:2217]: the setup for classic didn't actually change (submit first tx, such that available balance left is 1, then submit another tx and fail due to insufficient balance), hence I'm not expecting any new assertions to fail. a new setup for soroban was added (to produce the same failures but with enough balance to cover just the inclusion fee) but the expected behavior is the same, hence those asserts didn't change

[the remark this answers is not in the document]
Comment (dmkozh) [reply src/herder/test/TransactionQueueTests.cpp:2217]: Hmm, I thought Soroban setup has already been there before this PR, which is why I got confused by this change.  If Soroban scenario hasn't been covered before, then this makes sense to me.

File: src/test/TxTests.cpp [hunk 8]
```diff
@@ -1664,13 +1671,15 @@ static TransactionEnvelope
 sorobanEnvelopeFromOps(Hash const& networkID, TestAccount& source,
                        std::vector<Operation> const& ops,
                        std::vector<SecretKey> const& opKeys,
-                       SorobanResources const& resources, uint32_t totalFee,
-                       uint32_t resourceFee, std::optional<std::string> memo,
+                       SorobanResources const& resources, int64_t totalFee,
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/test/TxTests.cpp:1674]: ok, updated

File: src/herder/test/TransactionQueueTests.cpp [hunk 9]
```diff
@@ -1184,11 +1181,42 @@ TEST_CASE("Soroban TransactionQueue limits",
...
+        op0.body.type(INVOKE_HOST_FUNCTION);
+        auto& ihf0 = op0.body.invokeHostFunctionOp().hostFunction;
+        ihf0.type(HOST_FUNCTION_TYPE_CREATE_CONTRACT);
+
+        auto tx = transactionFrameFromOps(app->getNetworkID(), root, {op0}, {});
+        REQUIRE(app->getHerder().recvTransaction(tx, false) ==
+                TransactionQueue::AddResult::ADD_STATUS_ERROR);
+        REQUIRE(tx->getResult().result.code() == txMALFORMED);
+    }
     SECTION("negative inclusion fee")
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/test/TransactionQueueTests.cpp:1196]: added

File: src/herder/test/TransactionQueueTests.cpp [hunk 10]
```diff
@@ -1184,11 +1181,42 @@ TEST_CASE("Soroban TransactionQueue limits",
...
+        REQUIRE(fb->getFullFee() < fb->declaredSorobanResourceFee());
         REQUIRE(app->getHerder().recvTransaction(fb, false) ==
                 TransactionQueue::AddResult::ADD_STATUS_ERROR);
+        REQUIRE(fb->getResultCode() ==
+                TransactionResultCode::txSOROBAN_INVALID);
+    }
+    SECTION("negative declared resource fee")
+    {
+        auto invalid = createUploadWasmTx(*app, account1, initialInclusionFee,
+                                          -1, resAdjusted);
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/test/TransactionQueueTests.cpp:1210]: Done.

File: src/test/TxTests.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/test/TxTests.cpp:1674]: Updated.

File: src/herder/test/TransactionQueueTests.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/test/TransactionQueueTests.cpp:2217]: oh, I see the confusion now. there were soroban scenarios indeed, but they were too strict, as they didn't provide enough balance to cover the inclusion fee. The updates introduced new scenarios, and changed the setup to cover the inclusion fee, but fail to cover the full fee (to test for a bug we fixed recently). The failure mode is the same (insufficient fee), which is why the test cases themselves didn't change.

File: src/herder/TransactionQueue.cpp [hunk 11]
```diff
@@ -396,8 +398,8 @@ TransactionQueue::canAdd(TransactionFrameBasePtr tx,
     int64_t totalFees = feeStateIter == mAccountStates.end()
                             ? 0
                             : feeStateIter->second.mTotalFees;
-    if (getAvailableBalance(ltx.loadHeader(), feeSource) - newFullFee <
-        totalFees)
+    if (getAvailableBalance(ltx.loadHeader(), feeSource) <
+        totalFees + newFullFee)
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/TransactionQueue.cpp:402]: So we actually don't get to this line, because we fail earlier in `checkValid` due to account insufficient balance to cover `newFullFee`. To get here, we either need an int64_max balance, or a bunch of txs in the queue that add up to int64_max. I'll just keep this logic like in master (without moving newFullFee to RHS) for now

**Comment by dmkozh** [comment]
r+ 968cb21

**Comment by dmkozh** [comment]
r+ a366f61
